New job for mass registration of container hosts to Satellite

sthirugn opened the pull request:
Background: We need a Jenkins job to quickly test mass content host registrations to Satellite.

This new job will do the following:
1. (Optionally) setup the container host by installing ansible and setup all the necessary repos for running the script
2. Run mass registration script accepting variable parameters from the user to customize the specific needs of a test scenario.

The idea behind this specific job for now is automate the manual steps in Jenkins. Specific improvement around the playbook used and the flood.py script are planned for the near future.

JacobCallahan commented on jobs/satellite6-mass-registration.yaml:
```diff
@@ -0,0 +1,41 @@
+            name: Exit_CRITERIA
+            choices:
+                - 'register'
+                - 'katello-agent'
```
how about adding a time in seconds option? The flood script already supports this. The behavior is how many seconds the host should be kept alive after installing katello-agent.

sthirugn replied:
Good point, updated now

JacobCallahan commented on scripts/satellite6-mass-registration.sh:
```diff
@@ -0,0 +1,27 @@
+ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@"${CONTAINER_HOST}" << EOF
+  echo "Run the command for the container hosts registration"
+  cd /root/content-host-d
+  python flood.py rhel74 "${SATELLITE_HOST}" "${CONTENT_HOST_PREFIX}" "${ACTIVATION_KEY}" "${NUMBER_OF_HOSTS}" --exit-criteria "${EXIT_CRITERIA}"
```
This command is going to change soon. The new version of the master script will be much more like the guest branch, in that is does away with positional arguments.

sthirugn replied:
ACK, we can update this once flood.py is updated.

sthirugn commented:
@JacobCallahan updated to use the latest flood.py and is ready for final review/merge

JacobCallahan commented:
@sthirugn do you want to enforce activation keys, or let people specify none / organization / environment/ etc? If the first, then we're good.

sthirugn commented:
@JacobCallahan so far activation keys were used for all scenario testing.  I would vote to get this in with mandatory activation key and improve other options later as needed.

ehelms commented on jobs/satellite6-mass-registration.yaml:
```diff
@@ -0,0 +1,39 @@
+- job:
+    disabled: false
```
This is implied behavior if you want to drop it.

sthirugn replied:
Done!

ehelms commented:
Some general thoughts, take them with a grain of salt from our experiences.

 1) You could consider writing this as a Pipeline job the so called wave of the future
 2) You could use Ansible for nearly the entire process given you are ssh'ing into a machine and then running some Ansible. And Ansible is all about sshing into a machine and running X.
 3) One thing we've strived to do is try to make things easy to run outside of Jenkins, the bash script in theory does that except that all of the environment variable parameters make that more difficult for someone to take this script and construct. With bash, I'm not sure thats easier to achieve outside of maybe creating a properties file to define all variables in. If you considered Ansible, then its simply define these variables and run this playbook.

Another way to think about #3 is, treat your logic like you would create a tool for a user to run, then simply use Jenkins as an async task runner that calls that same logic.

evgeni commented on scripts/satellite6-mass-registration.sh:
```diff
@@ -0,0 +1,32 @@
+# Install python dependencies
+echo "Sourcing config files for cdn registration.."
+source ${CONFIG_FILES}
```
where is this variable comming from?

sthirugn replied:
I missed adding - config-file-provider: fixed locally, will update shortly which will provide this variable.

evgeni commented on scripts/satellite6-mass-registration.sh:
```diff
@@ -0,0 +1,32 @@
+# Install python dependencies
+echo "Sourcing config files for cdn registration.."
+source ${CONFIG_FILES}
+source config/subscription_config.conf
```
I guess this file has to be deployed by hand, as it will contain the `RHN_*` variables?

sthirugn replied:
I missed adding - config-file-provider: fixed locally, will update shortly

evgeni commented on jobs/satellite6-mass-registration.yaml:
```diff
@@ -0,0 +1,38 @@
+            default: false
+            description: If you already had setup the container host, make sure /root/content-host-d folder has the flood.py script.
+    builders:
+        - shining-panda:
```
does it need to be inside a venv?

sthirugn replied:
I put it inside a venv to give room for future development in python.

sthirugn commented:
@ehelms  and @evgeni ty for the review.  This job was initially intended to be a simple wrapper for the script.  But I totally agree that we should leverage more of ansible instead of direct ssh-ing to the slave.  I also like the idea that the playbooks should be capable of running by themselves or in jenkins so anyone can run it easily outside of jenkins. The code changes are made locally, will update the PR here shortly.

sthirugn commented:
thank you @evgeni and @ehelms.  I moved all the logic to two playbooks which can be invoked in jenkins or by themselves.  Ready for review.

Also I did the following:
- for the builder, I removed venv for now and we can add it later as needed
- updated config-file-provider to pull in data from jenkins CONFIG_FILES
- Tested with a similar job and it works great

JacobCallahan commented:
External PR has been merged!
https://github.com/JacobCallahan/content-host-d/pull/11

evgeni commented on scripts/satellite6-mass-registration.sh:
```diff
@@ -0,0 +1,15 @@
+echo "Setting up Container Host"
+if [ "${SETUP_CONTAINER_HOST}" == "true" ]; then
+    cd playbooks
+    sed -i "s/^localhost.*/${CONTAINER_HOST}/" hosts
```
instead of doing this, you can also just do `ansible-playbook --inventory=${CONTAINER_HOST}, …` (the comma at the end is important!). the only change needed to your playbooks is `hosts: all` instead of `hosts: home` :)

sthirugn replied:
good idea, updated now.

evgeni commented on scripts/satellite6-mass-registration.sh:
```diff
@@ -0,0 +1,15 @@
+    cd playbooks
+    sed -i "s/^localhost.*/${CONTAINER_HOST}/" hosts
+    # Note that the following playbook is temporarily hardcoded to generate rhel74 image.
+    ansible-playbook --extra-vars "RHN_USERNAME=${RHN_USERNAME} RHN_PASSWORD=${RHN_PASSWORD} RHN_POOLID=${RHN_POOLID} WORKSPACE=${WORKSPACE}" chd-setup.yaml
```
this will break if `RHN_PASSWORD` (or any other var) has special chars in it (like ` `, `!`, `$` etc).

using a pipeline with `[ansiblePlaybook](https://github.com/jenkinsci/ansible-plugin#extra-variables)` would solve that for you *hint* *hint* ;)

sthirugn replied:
the username has `@` and it worked great, do you mean it would fail for specific special characters?

evgeni replied:
Yeah, anything that is special to the shell running it, like the examples I gave above.

sthirugn replied:
I created a new issue https://github.com/SatelliteQE/robottelo-ci/issues/1035 to handle this since the current values of RHN username/password/poolid work fine with this implementation. I propose to get this PR merged so folks can start using them and meanwhile working on fixing handling special characters in shell.

evgeni replied:
works for me :)

evgeni commented:
@sthirugn that looks much better now! :)

sthirugn commented:
@evgeni thanks, your new comments were updated/responded.

JacobCallahan commented:
It's an ACK from my end.

sthirugn commented:
@ehelms gave an approval via irc

sthirugn commented:
thank you @JacobCallahan @ehelms and @evgeni for all valuable feedbacks.
